### PR TITLE
feat: add purchase ticket endpoint

### PR DESCRIPTION
**Pull request by MonicaTanios**
Co-authored-by: Michael Safwat <michaelsafwat.hanna@gmail.com>

### CODE REVIEW

File: src/tickets/dto/purchase-ticket.dto.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { BaseTicket } from '../interfaces/ticket.interface';
+
+export class PurchaseTicketDto implements BaseTicket {
+  userId: string;
+  quantity: number;
+  price: number;
```
Comment (yousinix) [inline src/tickets/dto/purchase-ticket.dto.ts:6]: I don't think we need to re-declare those properties, they are already inherited from `BaseTicket`

File: src/tickets/tickets.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { Test } from '@nestjs/testing';
+import { PurchaseTicketDto } from './dto/purchase-ticket.dto';
+import { TicketsService } from './tickets.service';
+import { TotpService } from './totp.service';
+
+describe('TicketsController', () => {
```
Comment (yousinix) [inline src/tickets/tickets.controller.spec.ts]: @MichaelSafwatHanna can you check this one, to meet the latest changes as the `user` module?

File: src/tickets/tickets.service.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+@Injectable()
+export class TicketsService {
+  constructor(private readonly totpService: TotpService) {}
+  purchase(purchaseTicketDto: PurchaseTicketDto): Ticket {
+    const user: User = {
+      id: purchaseTicketDto.userId,
+      secret: 'WKSFANU2PIIOF5GK',
+      email: 'zerobasedteam@gmail.com',
+      name: 'zero-based',
+    };
```
Comment (yousinix) [inline src/tickets/tickets.service.ts]: We should get the user from `UserService`, we already have a single declared user there.

File: src/tickets/totp.service.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { Injectable } from '@nestjs/common';
+import { TOTP } from '@otplib/core';
+import { totp } from '@otplib/preset-default';
+
+@Injectable()
+export class TotpService extends TOTP {
```
Comment (yousinix) [inline src/tickets/totp.service.ts]: I don't know if we really need to create this as a service, we won't test it or mock it as it is already an external code. We can simply call the `otp` library directly.
What do you think @MonicaTanios @MichaelSafwatHanna?

File: src/tickets/tickets.service.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    const isValid = this.totpService.check(purchaseTicketDto.totp, user.secret);
+    if (isValid) {
+      const transaction: Transaction = {
+        id: '1',
+        userId: '1',
```
Comment (yousinix) [inline src/tickets/tickets.service.ts]: Use `user.id`

File: src/tickets/tickets.service.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        amount: 6,
+        referenceId: '3',
```
Comment (yousinix) [inline src/tickets/tickets.service.ts]: Should be `null` in case of purchase.

File: src/tickets/tickets.service.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      email: 'zerobasedteam@gmail.com',
+      name: 'zero-based',
+    };
+
+    const isValid = this.totpService.check(purchaseTicketDto.totp, user.secret);
+    if (isValid) {
+      const transaction: Transaction = {
+        id: '1',
+        userId: '1',
+        amount: 6,
```
Comment (yousinix) [inline src/tickets/tickets.service.ts]: Use `purchaseTicketDto.quanitity * purchaseTicketDto.price`

File: src/tickets/tickets.service.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        referenceId: '3',
+      };
+      const ticket: Ticket = {
+        id: '1',
+        userId: '1',
+        quantity: 2,
+        price: 3,
```
Comment (yousinix) [inline src/tickets/tickets.service.ts]: Use values from predefined variables as above.

**Review by yousinix — CHANGES_REQUESTED** [review]

File: src/tickets/totp.service.ts (hunk 4 above)
Comment (MichaelSafwatHanna) [reply src/tickets/totp.service.ts]: We figured we can override it to change the algorithm default options like the window for example to not make those changes inside the `TicketService`, maybe we can discuss a more simple solution.

File: src/tickets/tickets.service.ts (hunk 3 above)
Comment (MichaelSafwatHanna) [reply src/tickets/tickets.service.ts]: But in the future, we'll make a database call here to get the user directly as the `TicketService` does not depend on the `UserService`, what do you think?

Comment (yousinix) [reply src/tickets/tickets.service.ts]: `UserService` should be the one doing the database calls on behalf of the `TicketService`.

File: src/tickets/dto/purchase-ticket.dto.ts (hunk 1 above)
Comment (MichaelSafwatHanna) [reply src/tickets/dto/purchase-ticket.dto.ts:6]: They're implemented from the interface as we need the DTO to be a class to be validated after enabling the validation on requests

File: src/tickets/totp.service.ts (hunk 4 above)
Comment (MichaelSafwatHanna) [reply src/tickets/totp.service.ts]: Agreed on removing it and calling the `otp` library directly.

**Review by yousinix — APPROVED** [review]

**Pull request merged**